File: app/src/main/java/com/maple/customview/view/DragView.java
package com.maple.customview.view;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

/**
 * 随手指移动的view
 */
public class DragView extends View {

    public DragView(Context context) {
        super(context);
    }

    public DragView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public DragView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    int lastX = 0;
    int lastY = 0;

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        // 获取手指触摸点的横纵坐标
        int x = (int) event.getX();
        int y = (int) event.getY();
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                // 开始位置
                lastX = x;
                lastY = y;
                break;
            case MotionEvent.ACTION_MOVE:
                // 计算移动距离
                int offsetX = x - lastX;
                int offsetY = y - lastY;
                // 调用layout方法，重新放置
                layout(getLeft() + offsetX, getTop() + offsetY,
                        getRight() + offsetX, getBottom() + offsetY);
                break;
            case MotionEvent.ACTION_UP:
                break;
            case MotionEvent.ACTION_CANCEL:
                break;
        }
        return true;
    }

}

File: app/src/main/java/com/maple/customview/view/DragView.kt
package com.maple.customview.view

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

/**
 * 随手指移动的view
 */
class DragView : View {

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    private var lastX = 0F
    private var lastY = 0F

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                // 获取手指触摸点的横纵坐标,设置开始位置
                lastX = event.x
                lastY = event.y
            }
            MotionEvent.ACTION_MOVE -> {
                // 计算移动距离
                val offsetX = (event.x - lastX).toInt()
                val offsetY = (event.y - lastY).toInt()
                // 调用layout方法，重新放置
                // layout(left + offsetX, top + offsetY, right + offsetX, bottom + offsetY)
                // 另一种写法
                offsetLeftAndRight(offsetX)
                offsetTopAndBottom(offsetY)
            }
            MotionEvent.ACTION_UP -> {
            }
            MotionEvent.ACTION_CANCEL -> {
            }
        }
        return true
    }

}
